docs/design/signature-provider-plugins.md

```diff
@@ -0,0 +1,444 @@
+               TargetPath="%(ReferenceCopyLocalPaths.DestinationSubPath)" />
+       </ItemGroup>
+   </Target>
+   
+   <!-- This target adds the generated deps.json file to the build output directory. -->
+   <Target Name="AddBuildDependencyFileToBuiltProjectOutputGroupOutput"
+           BeforeTargets="BuiltProjectOutputGroup"
+           Condition=" '$(GenerateDependencyFile)' == 'true'">
+   
+       <ItemGroup>
+           <BuiltProjectOutputGroupOutput
+               Include="$(ProjectDepsFilePath)"
+               TargetPath="$(ProjectDepsFileName)"
+               FinalOutputPath="$(ProjectDepsFilePath)" />
+       </ItemGroup>
+   </Target>
+   ```
+
+#### <a name="plugin-json-file"></a>The `plugin.json` file
+
+Sign CLI needs to load and execute plugins.  The general problem is that Sign CLI needs to know which assemblies in a plugin to load, which types to instantiate, how to initialize those objects, and so forth.  To simplify matters, plugins will embed this information in a `plugin.json` JSON file in their package's root directory.  The file should include the following properties:
+
+* `name`: The plugin's command name (e.g.:  `azure-key-vault` in `sign code azure-key-vault`).
+* `description`: The plugin's command descripton, to be displayed in command help.
+* `entryPoint`: Information for plugin instantiation.
+* `filePath`: The full file path within the package, relative to the package's root directory, for the assembly which contains the public implementation of an interface defined in the interfaces package.
+    * The path must be in its simplest form, without `..` or `.` directories.
+    * The directory separator must be `/`.
+    * The path must not have a leading slash `/`.
+    * The path must be case-sensitive.
+* `implementationTypeName`: The fully qualified type name of the public type implementing the public interface defined in the interfaces package.
+* `interfaceTypeName`: The fully qualified type name of the public type in the interfaces package that `implementationTypeName` implements.
+* `parameters`: The plugin's command options.
+    * `name`: The option name.  Not displayed.
+    * `description`: The option's description.  Displayed in command help.
+    * `aliases`: The option's names.  Should include both long and short forms, (e.g.:  `--azure-key-vault-managed-identity` and `-kvm`, respectively).  A user will type these option names.
+    * `dataType`: The option's data type.  Lets Sign CLI know how to parse user input.
+    * `defaultValue`: The option's default value.  Only used if the effective value for `isRequired` is `false`; otherwise, this is ignored.
+    * `isRequired`: Whether the option is required or not.
+
+Parameter | Required | JSON Type | Default | Possible Values
+-- | -- | -- | -- | --
+`name` | yes | string | N/A | N/A
+`description` | yes | string | N/A | N/A
+`aliases` | yes | array of strings | N/A | N/A
+`dataType` | no | string | `Text` | `Text`, `Boolean`, `Uri`
```

Comment:
I imagine, plugin would have to validate parameters anyway, so why not just read and pass them to a plugin as strings, then let it sort out the conversion/interpretation?

Reply:
Thoughts, @clairernovotny?

Reply:
I don't have a strong opinion on this other than having common parameters validated in the base could provide a more consistent experience.